File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        
        if (s.isEmpty()) {
            return "";
        }
        
        int[] ind = {0, 0};
        
        for(int i=0; i<s.length(); i++) {
            int[] ind1 = expand(s, i, i);
            if (ind1[1] - ind1[0] > ind[1] - ind[0]) {
                ind = ind1;
            }
            if (i+1 < s.length() && s.charAt(i) == s.charAt(i+1)) {
                int[] ind2 = expand(s, i, i+1);
                if (ind2[1] - ind2[0] > ind[1] - ind[0]) {
                    ind = ind2;
                }
            }
        }
        
        return s.substring(ind[0], ind[1] + 1);
            
        
    }
    
    private int[] expand(String s, int i, int j) {
        for(; i>=0 && j<s.length(); i--, j++) {
            if (s.charAt(i) != s.charAt(j)) break;
        }
        return new int[] {i+1, j-1};
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        
        if not s:
            return ""
        
        left = 0
        right = 0
        
        for i in range(len(s)):
            len1 = self.expand(s, i, i)
            len2 = self.expand(s, i, i+1)
            result = max(len1, len2)
            
            if result > right - left + 1:
                left = i - (result-1)//2
                right = i + result//2
        
        return s[int(left):int(right+1)]
        
        
    def expand(self, word, left, right):
        while 0 <= left and right < len(word) and word[left] == word[right]:
            left -= 1
            right += 1

        return right - left - 1
        